### PR TITLE
Get remote user name during account creation or update taking into ac…

### PR DESCRIPTION
**Pull request by davivel**
…count a possible permanent redirection in the WebDAV entry point.

### CODE REVIEW

File: src/com/owncloud/android/authentication/AuthenticatorAsyncTask.java [hunk 1]
```diff
@@ -72,18 +72,32 @@ protected RemoteOperationResult doInBackground(Object... params) {
             );
             result = operation.execute(client);
 
+            String permanentLocation = null;
```
Comment (davigonz) [inline src/com/owncloud/android/authentication/AuthenticatorAsyncTask.java]: Maybe a comment here could help to understand the use of this variable

Comment (davigonz) [inline src/com/owncloud/android/authentication/AuthenticatorAsyncTask.java]: Maybe a comment here could help to understand the use of this variable. What do you mean with permanent location?

File: src/com/owncloud/android/authentication/AuthenticatorAsyncTask.java [hunk 2]
```diff
@@ -72,18 +72,32 @@ protected RemoteOperationResult doInBackground(Object... params) {
...
+                if (permanentLocation == null) {
+                    GetRemoteUserInfoOperation remoteUserNameOperation = new GetRemoteUserInfoOperation();
+                    result = remoteUserNameOperation.execute(client);
+                } else {
+                    // we can't assume that any subpath of the domain is correctly redirected; ugly stuff
+                    GetRemoteUserInfoOperation remoteUserNameOperation = new GetRemoteUserInfoOperation();
+                    OwnCloudClient redirectedClient = OwnCloudClientFactory.createOwnCloudClient(
+                        Uri.parse(AccountUtils.trimWebdavSuffix(permanentLocation)),
+                        mContext,
+                        true
```
Comment (davigonz) [inline src/com/owncloud/android/authentication/AuthenticatorAsyncTask.java]: It's a good practice to use variables when you use a boolean, in order to explain what that boolean is referring to

File: src/com/owncloud/android/authentication/AuthenticatorAsyncTask.java [hunk 3]
```diff
@@ -72,18 +72,32 @@ protected RemoteOperationResult doInBackground(Object... params) {
...
+                    );
+                    redirectedClient.setCredentials(credentials);
+                    result = remoteUserNameOperation.execute(redirectedClient);
+                }
             }
 
+            result.setLastPermanentLocation(permanentLocation);
```
Comment (davigonz) [inline src/com/owncloud/android/authentication/AuthenticatorAsyncTask.java]: Same here, a comment could help to understand better the code

File: src/com/owncloud/android/authentication/AuthenticatorAsyncTask.java (hunk 1 above)
Comment (davivel) [inline src/com/owncloud/android/authentication/AuthenticatorAsyncTask.java]: I see coming a new age of comments all over the code :'D :'D :'D

File: src/com/owncloud/android/authentication/AuthenticatorAsyncTask.java (hunk 2 above)
Comment (davivel) [inline src/com/owncloud/android/authentication/AuthenticatorAsyncTask.java]: With the cursor over the name of the method, click F1. The documentation of the method will be shown, including the name of the parameter and its meaning.

Declaring an assigning a variable to use it just once as a parameter seems too much IMHO.

File: src/com/owncloud/android/authentication/AuthenticatorAsyncTask.java [hunk 4]
```diff
@@ -72,18 +72,32 @@ protected RemoteOperationResult doInBackground(Object... params) {
...
-                String permanentLocation = redirectionPath.getLastPermanentLocation();
-                result.setLastPermanentLocation(permanentLocation);
+                permanentLocation = redirectionPath.getLastPermanentLocation();
             }
 
             // Operation - get display name
             if (result.isSuccess()) {
-                GetRemoteUserInfoOperation remoteUserNameOperation = new GetRemoteUserInfoOperation();
-                result = remoteUserNameOperation.execute(client);
+                if (permanentLocation == null) {
```
Comment (SamuAlfageme) [inline src/com/owncloud/android/authentication/AuthenticatorAsyncTask.java]: A bit more compact and DRY version of the code can be achieved by reference aliasing, which allows to remove one of the conditions and leave it as a single if:

```
if (result.isSuccess()) {
    GetRemoteUserInfoOperation remoteUserNameOperation = new GetRemoteUserInfoOperation();
    if (permanentLocation != null) {
        // we can't assume that any subpath of the domain is correctly redirected; ugly stuff
        OwnCloudClient redirectedClient = OwnCloudClientFactory.createOwnCloudClient(
            Uri.parse(AccountUtils.trimWebdavSuffix(permanentLocation)),
            mContext,
            true
        );
        redirectedClient.setCredentials(credentials);
        client = redirectedClient;
    }
    result = remoteUserNameOperation.execute(client);
}
```

**Comment by davivel** [comment]
Updated after CRs.

Please, @SamuAlfageme , @davigonz , check again.

**Comment by davigonz** [comment]
:+1:

**Comment by SamuAlfageme** [comment]
Lookin' good!
- [x] Account creation: follows the `301` like a charm. 
- [x] Account update: even updates the URL endpoint for the one of the redirection. 

So it's a 👍

**Comment by davivel** [comment]
💃

**Pull request merged**